compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3403,146 +3403,434 @@ TR::Register *OMR::X86::TreeEvaluator::encodeUTF16Evaluator(TR::Node *node, TR::
    }
 #endif
 
-TR::Register *OMR::X86::TreeEvaluator::arraysetEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+static void packUsingShift(TR::Node* node, TR::Register* tempReg, TR::Register* sourceReg, int32_t size, TR::CodeGenerator* cg)
    {
-   // arrayset
-   //    address
-   //    fill value
-   //    array size (in bytes, unsigned)
+   generateRegRegInstruction(MOV8RegReg, node, tempReg, sourceReg, cg);
+   generateRegImmInstruction(SHL8RegImm1, node, sourceReg, size*8, cg);
+   generateRegRegInstruction(OR8RegReg, node, sourceReg, tempReg, cg);
+   }
 
-   TR::Node* addressNode  = node->getChild(0); // Adress
-   TR::Node* valueNode = node->getChild(1);    // Value
-   TR::Node* sizeNode  = node->getChild(2);    // Size
+static void generateMovToMemInstructionsForArrayset(TR::Node* node, TR::Register* valueReg, uint8_t size, TR::Register* addressReg, int32_t index, TR::CodeGenerator* cg)
+   {
+   TR_X86OpCodes opcode;
+   switch(size)
+      {
+      case 1:
+         opcode = S1MemReg;
+         break;
+      case 2:
+         opcode = S2MemReg;
+         break;
+      case 4:
+         opcode = S4MemReg;
+         break;
+      case 8:
+         opcode = S8MemReg;
+         break;
+      case 16:
+         opcode = MOVDQUMemReg;
+         break;
+      }
+   generateMemRegInstruction(opcode, node, generateX86MemoryReference(addressReg, index, cg), valueReg, cg);
+   }
 
-   TR::Register* addressReg = TR::TreeEvaluator::intOrLongClobberEvaluate(addressNode, TR::Compiler->target.is64Bit(), cg);
-   TR::Register* valueReg = cg->evaluate(valueNode);
-   TR::Register* sizeReg = TR::TreeEvaluator::intOrLongClobberEvaluate(sizeNode, TR::TreeEvaluator::getNodeIs64Bit(sizeNode, cg), cg);
+static void packXMMWithMultipleValues(TR::Node* node, TR::Register* XMMReg, TR::Register* sourceReg, int8_t size, TR::CodeGenerator* cg)
+   {
+   switch(size)
+      {
+      case 1:
+      case 2:
+         {
+         uint8_t MASKOFSIZEONE[] =
+            {
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            };
+
+         uint8_t MASKOFSIZETWO[] =
+            {
+            0x00, 0x01, 0x00, 0x01,
+            0x00, 0x01, 0x00, 0x01,
+            0x00, 0x01, 0x00, 0x01,
+            0x00, 0x01, 0x00, 0x01,
+            };
```

Comment:
Minor performance point:  I think it would be better to mark these `static` to avoid the auto array initialization each time through here.